src/Microsoft.AspNetCore.SignalR.Client/RpcConnection.cs

```diff
@@ -0,0 +1,147 @@
+        public static Task<RpcConnection> ConnectAsync(Uri url, IInvocationAdapter adapter, ITransport transport, HttpClient httpClient, PipelineFactory pipelineFactory) => ConnectAsync(url, transport, httpClient, pipelineFactory, NullLoggerFactory.Instance);
+
+        public static async Task<RpcConnection> ConnectAsync(Uri url, IInvocationAdapter adapter, ITransport transport, HttpClient httpClient, PipelineFactory pipelineFactory, ILoggerFactory loggerFactory)
+        {
+            // Connect the underlying connection
+            var connection = await Connection.ConnectAsync(url, transport, httpClient, pipelineFactory, loggerFactory);
+
+            // Create the RPC connection wrapper
+            return new RpcConnection(connection, adapter);
+        }
+
+        private async Task ReceiveMessages(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                // This is a little odd... we want to remove the InvocationRequest once and only once so we pull it out in the callback,
+                // and stash it here because we know the callback will have finished before the end of the await.
+                InvocationRequest irq = default(InvocationRequest);
+                var result = await _adapter.ReadInvocationResultDescriptorAsync(_stream, id =>
+                {
+                    if (!_pendingCalls.TryRemove(id, out irq))
+                    {
+                        throw new InvalidOperationException($"Unsolicited response received for invocation '{id}'");
+                    }
+                    return irq.ResultType;
+                }, cancellationToken);
+
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    break;
+                }
+
+                Debug.Assert(irq.Completion != null, "Didn't properly capture InvocationRequest in callback for ReadInvocationResultDescriptorAsync");
+
+                // If the invocation hasn't been cancelled, dispatch the result
+                if (!irq.CancellationToken.IsCancellationRequested)
+                {
+                    irq.Registration.Dispose();
+
+                    // Complete the request based on the result
+                    if (!string.IsNullOrEmpty(result.Error))
+                    {
+                        irq.Completion.TrySetException(new Exception(result.Error));
+                    }
+                    else
+                    {
+                        irq.Completion.TrySetResult(result.Result);
+                    }
+                }
+            }
+        }
+
+        private string GetNextId()
+        {
+            // Increment returns the incremented value, so we subtract one to get the previous value.
+            return (Interlocked.Increment(ref _nextId) - 1).ToString();
```

Comment:
or you could initialize _nextId to -1. Starting at 1 is fine too - this is an opaque identifier